### DISCUSSION TITLE
How does dotnet know the test framework?

### DISCUSSION THREAD

**Discussion by ShuiRuTian** [xunit/xunit#2171] [General]
`dotnet new xunit` could create a project. Then `dotnet test` could run test cases.

In my opinion, xunit is only a package like others, so how does dotnet find that xunit is a test framework and then start it? It looks like something magic!

**Comment by bradwilson** [comment]
The VS Test engine (which we plug into) finds test adapters by looking for NuGet package references that contain assemblies whose name ends in `.testadapter.dll` and it exposes types that implement the discovery and execution interfaces. It then dynamically loads that test adapter, and asks it to perform discovery on the assembly.

It even supports multiple test adapters in the same project. We validate that this works in our integration tests: https://github.com/xunit/xunit.integration/tree/main/test-explorer
